src/Init/Data/String/Basic.lean

```diff
@@ -196,16 +196,48 @@ def utf8PrevAux : List Char → Pos → Pos → Pos
     let i' := i + c
     if i' = p then i else utf8PrevAux cs i' p
 
+/--
+Returns the position in a string before a specified position, `p`. If `p = ⟨0⟩`, returns `0`.
+If `p` is not a valid position, the result is unspecified.
+
+Examples:
+* `"abc".prev ⟨2⟩ = String.Pos.mk 1`
+* `"abc".prev ⟨0⟩ = String.Pos.mk 0`
+* `"L∃∀N".prev ⟨4⟩ = String.Pos.mk 1`, since `'∃'` is a multi-byte UTF-8 character
```

Comment:
These examples all worry me, because they seem to suggest that users should be counting bytes and creating a `String.Pos` from scratch. Additionally, the use of two separate notations for `String.Pos.mk` risks throwing off some users, and I think the `String` API is more likely to attract new users than many other areas. Finally, it shows off an unspecified result, which I think is better left unwritten.

I also think that byte 4 is valid, based on:
```
#eval let i := ⟨0⟩; let str := "L∃∀N"; str.next i
```
yielding 1 and
```
#eval let i := ⟨0⟩; let str := "L∃∀N"; str.next <| str.next i
```
yielding 4.

What about:
```suggestion
Examples:
* `"abc".prev ⟨2⟩ = ⟨1⟩`
* `"abc".prev ⟨0⟩ = ⟨0⟩`
* `"L∃∀N".prev ⟨3⟩` is unspecified, since `'∃'` is a multi-byte UTF-8 character and byte 3 is in the middle of it
```

Reply:
> I also think that byte 4 is valid, based on:

 I intended for `"L∃∀N".prev ⟨4⟩ = String.Pos.mk 1` to illustrate a valid result in the case where the string contains multi-byte characters. I didn't include an unspecified example here (unlike [`String.next`](https://github.com/leanprover/lean4/blob/master/src/Init/Data/String/Basic.lean#L184)) but can definitely add one.

More generally, I think your concerns regarding byte counts from scratch is spot on. What about the following modification modeled after your suggestion for `String.atEnd` to encourage more natural usage:

```suggestion
Examples:
Given `def abc := "abc"` and `def lean := "L∃∀N"`,
* `abc.endPos |> abc.prev = ⟨2⟩`
* `lean.endPos |> lean.prev |> lean.prev |> lean.prev = ⟨1⟩`
* `"L∃∀N".prev ⟨3⟩` is unspecified, since `'∃'` is a multi-byte UTF-8 character and byte 3 is in the middle of it
```

The other advantage here is demonstrating that if you call `String.prev` with a constructed `String.pos`, instead of with an iterative pattern, you run the risk of unspecified results.

Here is an alternative version with `String.get` which is more explicit at the cost of complicating the example code:
```suggestion
Examples:
Given `def abc := "abc"` and `def lean := "L∃∀N"`,
* `abc.get <| abc.endPos |> abc.prev = 'c'`
* `lean.get <| lean.endPos |> lean.prev |> lean.prev |> lean.prev = '∃'`
* `"L∃∀N".prev ⟨3⟩` is unspecified, since `'∃'` is a multi-byte UTF-8 character and byte 3 is in the middle of it
```

What do you think?

Reply:
I like it, except for the mixing of `<|` and `|>` - the precedence of them is not necessarily obvious. What about parens around the chains of `|>` instead?

Reply:
:+1: I changed these examples to use parens instead of `<|`.

Additionally, I converted the examples for the similar `String.next` which was documented a few weeks ago in #4001 to this iterative style instead of constructing positions.

Reply:
This looks great, thanks!